### PR TITLE
test: Adding testing framework along with unit tests for the recipes

### PR DESCRIPTION
**Pull request by rtpa25**
## Summary of change
- Set up the test environment 
- added config tests
- added code fixes

## Related issues

-   https://github.com/supertokens/supertokens-golang/issues/24

## Test Plan

(Write your test plan here. If you changed any code, please provide us with clear instructions on how you verified your changes work. Bonus points for screenshots and videos!)

## Documentation changes

(If relevant, please create a PR in our [docs repo](https://github.com/supertokens/docs), or create a checklist here highlighting the necessary changes)

## Checklist for important updates

-   [ ] Changelog has been updated
-   [ ] `coreDriverInterfaceSupported.json` file has been updated (if needed)
    -   Along with the associated array in `supertokens/constants.go`
-   [ ] `frontendDriverInterfaceSupported.json` file has been updated (if needed)
-   [ ] Changes to the version if needed
    -   In `supertokens/constants.go > version variable`
-   [x] Had installed and ran the pre-commit hook
-   [x] Issue this PR against the latest non-released version branch.
    -   To know which one it is, run find the latest released tag (`git tag`) in the format `vX.Y.Z`, and then find the latest branch (`git branch --all`) whose `X.Y` is greater than the latest released tag.
    -   If no such branch exists, then create one from the latest released branch.

## Remaining TODOs for this PR

- [ ] Finish basic unit tests for all recipes
   - Calling recipe functions and making sure they work
   - Different config values and result from that
   - Test for querier
- [x] Change to CICD to run these tests
- [x] Update the contributing guide
- [x] Run these tests against all versions of the core that are compatible with this to confirm that they are written correctly

### CODE REVIEW

File: supertokens/querier.go [hunk 1]
```diff
@@ -32,8 +32,8 @@ type Querier struct {
 
 var (
 	querierInitCalled     bool                  = false
-	querierHosts          []NormalisedURLDomain = nil
-	querierAPIKey         *string
+	QuerierHosts          []NormalisedURLDomain = nil //*these are changed
+	QuerierAPIKey         *string                     //*these are changed
```
Comment (rishabhpoddar) [inline supertokens/querier.go:36]: Don't add comments like this. Not needed.

File: supertokens/supertokens.go [hunk 2]
```diff
@@ -54,16 +54,21 @@ func supertokensInit(config TypeInput) error {
 	}
 
 	if config.Supertokens != nil {
-		hostList := strings.Split(config.Supertokens.ConnectionURI, ";")
-		var hosts []NormalisedURLDomain
-		for _, h := range hostList {
-			host, err := NewNormalisedURLDomain(h)
-			if err != nil {
-				return err
+		//*if pass an empty struct with no values inside that comes to this block
```
Comment (rishabhpoddar) [inline supertokens/supertokens.go:57]: Why the "*" in the comment?

File: supertokens/supertokens.go [hunk 3]
```diff
@@ -54,16 +54,21 @@ func supertokensInit(config TypeInput) error {
...
+				if err != nil {
+					return err
+				}
+				hosts = append(hosts, host)
 			}
-			hosts = append(hosts, host)
+			initQuerier(hosts, config.Supertokens.APIKey)
+		} else {
+			return errors.New("please provide 'ConnectionURI' value")
 		}
```
Comment (rishabhpoddar) [inline supertokens/supertokens.go:71]: Make this change a different PR

File: test/unittesting/utils.go [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+
+func killAllST() {
+	pids := getListOfPids()
+	for i := 0; i < len(pids); i++ {
+		stopST(pids[i])
+	}
+	resetAll()
+}
+
+func StartingHelper() {
```
Comment (rishabhpoddar) [inline test/unittesting/utils.go:164]: prefer if you call this `BeforeEach`

File: test/unittesting/utils.go [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+	killAllST()
+	setUpST()
+	startUpST("localhost", "8080")
+}
+
+func EndingHelper() {
```
Comment (rishabhpoddar) [inline test/unittesting/utils.go:170]: Prefer if you call this `AfterEach`

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: test/unittesting/utils.go [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+		}
+		result = append(result, string(data))
+	}
+	return result
+}
+
+//*this copies the config.yaml from temp folder of the root and puts it into the root level.
+func setUpST() {
+	slashesNeededToGoUp := returnNumberOfDirsToGoUpFromCurrentWorkingDir()
+	os.Setenv("INSTALL_PATH", slashesNeededToGoUp+"supertokens-root")
```
Comment (rishabhpoddar) [inline test/unittesting/utils.go:64]: Don't set `INSTALL_PATH` as an env var. Make the user pass it and then you figure out how many `../` needs to be prepended to it.

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: recipe/emailpassword/api/emailExists.go [hunk 7]
```diff
@@ -16,15 +16,20 @@
...
 
 import (
+	"fmt"
+
 	"github.com/supertokens/supertokens-golang/recipe/emailpassword/epmodels"
 	"github.com/supertokens/supertokens-golang/supertokens"
 )
 
 func EmailExists(apiImplementation epmodels.APIInterface, options epmodels.APIOptions) error {
+	fmt.Println("this is api implementation", &apiImplementation.EmailExistsGET)
```
Comment (rishabhpoddar) [inline recipe/emailpassword/api/emailExists.go:26]: No print statements please. Make sure this doesn't happen again

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: recipe/emailpassword/recipeImplementation.go [hunk 8]
```diff
@@ -54,7 +54,7 @@ func MakeRecipeImplementation(querier supertokens.Querier) epmodels.RecipeInterf
 		}
 		status, ok := response["status"]
 		if ok && status.(string) == "OK" {
-			user, err := parseUser(response["user"])
+			user, err := ParseUser(response["user"])
```
Comment (rishabhpoddar) [inline recipe/emailpassword/recipeImplementation.go:57]: You should not expose an internal helper function like this.. If this is required in testing, copy the implementation of that in the test folder.

File: test/unittesting/config_test.go [hunk 9]
```diff
@@ -0,0 +1,1403 @@
...
+	if err != nil {
+		errorMessage := err.Error()
+		if errorMessage != "please provide at least one recipe to the supertokens.init function call" {
+			t.Errorf(err.Error())
+		} else {
+			assert.Equal(t, errorMessage, "please provide at least one recipe to the supertokens.init function call")
+		}
+	}
+
+	defer AfterEach()
```
Comment (rishabhpoddar) [inline test/unittesting/config_test.go:169]: If you use defer, you should put it right at the top of the function scope. In this case, you want it to be right after `StartUpST("localhost", "8080")`

File: test/unittesting/config_test.go [hunk 10]
```diff
@@ -0,0 +1,1403 @@
...
+			errMessage := err.Error()
+			if configValue.AppInfo.AppName != "SuperTokens" {
+				assert.Equal(t, errMessage, "Please provide your appName inside the appInfo object when calling supertokens.init")
+			} else if configValue.AppInfo.APIDomain != "api.supertokens.io" {
+				assert.Equal(t, errMessage, "Please provide your apiDomain inside the appInfo object when calling supertokens.init")
+			} else if configValue.AppInfo.WebsiteDomain != "supertokens.io" {
+				assert.Equal(t, errMessage, "Please provide your websiteDomain inside the appInfo object when calling supertokens.init")
+			}
+		}
+		AfterEach()
```
Comment (rishabhpoddar) [inline test/unittesting/config_test.go:141]: Missing use of defer

File: test/unittesting/config_test.go [hunk 11]
```diff
@@ -0,0 +1,1403 @@
...
+
+		if configValue.AppInfo.APIBasePath != nil {
+			assert.Equal(t, "/test", supertokensInstance.AppInfo.APIBasePath.GetAsStringDangerous())
+			assert.Equal(t, "/test1", supertokensInstance.AppInfo.WebsiteBasePath.GetAsStringDangerous())
+		} else {
+			assert.Equal(t, "/auth", supertokensInstance.AppInfo.APIBasePath.GetAsStringDangerous())
+			assert.Equal(t, "/auth", supertokensInstance.AppInfo.WebsiteBasePath.GetAsStringDangerous())
+		}
+		AfterEach()
+	}
```
Comment (rishabhpoddar) [inline test/unittesting/config_test.go:85]: This type of testing is very weird.. You want to either separate them into different tests, or do init -> check -> init -> check.

Also, do you really need to start and stop ST in each loop here? Cause that will add more time to this test

File: test/unittesting/config_test.go [hunk 12]
```diff
@@ -0,0 +1,1403 @@
...
+	if err != nil {
+		if err != nil {
+			errorMessage := err.Error()
+			if errorMessage != `cookie same site must be one of "strict", "lax", or "none"` {
+				t.Errorf(err.Error())
+			} else {
+				assert.Equal(t, errorMessage, `cookie same site must be one of "strict", "lax", or "none"`)
+			}
+		}
+	}
```
Comment (rishabhpoddar) [inline test/unittesting/config_test.go:322]: This style of testing is very strange. Separate them into different tests, or then do init -> check -> init -> check

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: test/unittesting/config_test.go (hunk 10 above)
Comment (rishabhpoddar) [reply test/unittesting/config_test.go:141]: Check in all places.

File: test/unittesting/config_test.go [hunk 13]
```diff
@@ -0,0 +1,1403 @@
...
+		},
+		RecipeList: []supertokens.Recipe{},
+	}
+	BeforeEach()
+	StartUpST("localhost", "8080")
+	err := supertokens.Init(configValue)
+	if err != nil {
+		errorMessage := err.Error()
+		if errorMessage != "please provide at least one recipe to the supertokens.init function call" {
+			t.Errorf(err.Error())
```
Comment (rishabhpoddar) [inline test/unittesting/config_test.go:163]: Somewhere you are using Errorf, and somewhere just Error. Is this intentional or a mistake?

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: test/unittesting/utils.go [hunk 14]
```diff
@@ -0,0 +1,652 @@
...
+		}
+	}
+	return result
+}
+
+func getCurrTimeInMS() uint64 {
+	return uint64(time.Now().UnixNano() / 1000000)
+}
+
+func Shellout(command string) (string, string, error) {
```
Comment (rishabhpoddar) [inline test/unittesting/utils.go:116]: Why is this capital S?

File: test/unittesting/utils.go [hunk 15]
```diff
@@ -0,0 +1,652 @@
...
+	}
+	numberOfElems := len(arr) - counter
+	var dirUpSlash string
+	for i := 0; i < numberOfElems; i++ {
+		dirUpSlash += "../"
+	}
+	return dirUpSlash
+}
+
+func returnNumberOfSlashesRequiredToGoToRootOfTheProject() string {
```
Comment (rishabhpoddar) [inline test/unittesting/utils.go:236]: This seems to be redundant cause we already have `returnNumberOfDirsToGoUpFromCurrentWorkingDir`. Why is this here?

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: supertokens/utils.go [hunk 16]
```diff
@@ -104,13 +104,13 @@ func getLargestVersionFromIntersection(v1 []string, v2 []string) *string {
...
 	for i := 1; i < len(intersection); i++ {
-		maxVersionSoFar = maxVersion(intersection[i], maxVersionSoFar)
+		maxVersionSoFar = MaxVersion(intersection[i], maxVersionSoFar)
 	}
 	return &maxVersionSoFar
 }
 
 // MaxVersion returns max of v1 and v2
-func maxVersion(version1 string, version2 string) string {
+func MaxVersion(version1 string, version2 string) string {
```
Comment (rishabhpoddar) [inline supertokens/utils.go:113]: No need to make this capital. Instead, add a copy of this in the unittesting/testingutils.go file, and expose this via that and then use that in your tests.

File: recipe/session/config_test.go [hunk 17]
```diff
@@ -0,0 +1,1235 @@
...
+			Init(nil),
+		},
+	}
+	BeforeEach()
+	unittesting.StartUpST("localhost", "8080")
+	err := supertokens.Init(configValue)
+	if err != nil {
+		assert.Equal(t, err.Error(), "Please provide your apiDomain inside the appInfo object when calling supertokens.init")
+	}
+	AfterEach()
```
Comment (rishabhpoddar) [inline recipe/session/config_test.go:124]: Missing use of defer, and in incorrect place.

File: recipe/session/config_test.go [hunk 18]
```diff
@@ -0,0 +1,1235 @@
...
+		RecipeList: []supertokens.Recipe{
+			Init(nil),
+		},
+	}
+	BeforeEach()
+	unittesting.StartUpST("localhost", "8080")
+	err := supertokens.Init(configValue)
+	if err != nil {
+		assert.Equal(t, err.Error(), "Please provide your apiDomain inside the appInfo object when calling supertokens.init")
+	}
```
Comment (rishabhpoddar) [inline recipe/session/config_test.go:123]: If err is nill, then also we want to fail this test right? I think this type of checking has been ignored in other tests too. Please make sure to cover all the checks

File: recipe/emailpassword/override_test.go [hunk 19]
```diff
@@ -0,0 +1,392 @@
...
+	mux := http.NewServeMux()
+	mux.HandleFunc("/user", func(rw http.ResponseWriter, r *http.Request) {
+		userId := r.URL.Query().Get("userId")
+		fetchedUser, err := GetUserByID(userId)
+		if err != nil {
+			t.Error(err.Error())
+		}
+		jsonResp, err := json.Marshal(fetchedUser)
+		if err != nil {
+			t.Errorf("Error happened in JSON marshal. Err: %s", err)
```
Comment (rishabhpoddar) [inline recipe/emailpassword/override_test.go:100]: Is using Errorf intentional or a mistake? if it's a mistake, please be sure to remove it from all tests, else resolve this comment

File: test/unittesting/testingutils.go [hunk 20]
```diff
@@ -0,0 +1,624 @@
...
+		numberOfElems = len(arr) - counter
+	}
+	var dirUpSlash string
+	for i := 0; i < numberOfElems; i++ {
+		dirUpSlash += "../"
+	}
+	return dirUpSlash
+}
+
+func RemoveTrailingSlashFromTheEndOfString(input string) string {
```
Comment (rishabhpoddar) [inline test/unittesting/testingutils.go:230]: This function doesn't seem to be used anywhere. Remove it. Make sure to remove all unused functions?

File: test/unittesting/testingutils.go [hunk 21]
```diff
@@ -0,0 +1,624 @@
...
+	installationDir := os.Getenv("INSTALL_DIR")
+	if installationDir == "" {
+		installationDir = returnNumberOfDirsToGoUpFromCurrentWorkingDir(false) + "supertokens-root"
+	} else {
+		installationDir = returnNumberOfDirsToGoUpFromCurrentWorkingDir(true) + installationDir
+	}
+	return installationDir
+}
+
+func SetKeyValueInConfig(key string, value string) {
```
Comment (rishabhpoddar) [inline test/unittesting/testingutils.go:420]: This function is not being used anywhere. Shouldn't this be used in the tests?

File: test/unittesting/testingutils.go [hunk 22]
```diff
@@ -0,0 +1,624 @@
...
+	defer f.Close()
+
+	f.Write([]byte("\n" + key + ": " + value))
+
+	if _, err = f.WriteString(key + ": " + value + "\n"); err != nil {
+		panic(err)
+	}
+}
+
+func SetKeyAndNumberValueInConfig(key string, value int) {
```
Comment (rishabhpoddar) [inline test/unittesting/testingutils.go:437]: I don't think you need this function as in the config file, adding a number is just like adding a string. So you can replace calls to this with calls to `SetKeyValueInConfig` and convert the number to a string when calling the function?

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: recipe/session/testingUtils.go [hunk 23]
```diff
@@ -0,0 +1,38 @@
+/* Copyright (c) 2021, VRAI Labs and/or its affiliates. All rights reserved.
```
Comment (rishabhpoddar) [inline recipe/session/testingUtils.go:1]: Why can't all of these files be moved to tests folder and these functions exposed from that package?

File: supertokens/querier.go [hunk 24]
```diff
@@ -279,7 +279,7 @@ func (q *Querier) sendRequestHelper(path NormalisedURLPath, httpRequest httpRequ
 		return nil, readErr
 	}
 	if resp.StatusCode != 200 {
-		return nil, errors.New(fmt.Sprintf("SuperTokens core threw an error for a request to path: '%s' with status code: %v and message: %s", path.GetAsStringDangerous(), resp.StatusCode, body))
+		return nil, fmt.Errorf("SuperTokens core threw an error for a request to path: '%s' with status code: %v and message: %s", path.GetAsStringDangerous(), resp.StatusCode, body)
```
Comment (rishabhpoddar) [inline supertokens/querier.go:282]: Does this actually create a new error and return that? Or is this a mistake?

File: supertokens/utils.go [hunk 25]
```diff
@@ -153,6 +153,7 @@ func Send200Response(res http.ResponseWriter, responseJson interface{}) error {
 
 func SendNon200Response(res http.ResponseWriter, message string, statusCode int) error {
 	dw := MakeDoneWriter(res)
+
```
Comment (rishabhpoddar) [inline supertokens/utils.go:156]: Undo new line

File: recipe/emailpassword/api/emailExists.go [hunk 26]
```diff
@@ -25,6 +25,7 @@ func EmailExists(apiImplementation epmodels.APIInterface, options epmodels.APIOp
 		options.OtherHandler(options.Res, options.Req)
 		return nil
 	}
+
```
Comment (rishabhpoddar) [inline recipe/emailpassword/api/emailExists.go:28]: Undo new line

File: recipe/session/config_test.go (hunk 18 above)
Comment (rishabhpoddar) [reply recipe/session/config_test.go]: How have you made sure that this mistake is not there anywhere?

File: test/unittesting/testingutils.go (hunk 21 above)
Comment (rishabhpoddar) [reply test/unittesting/testingutils.go:434]: @rtpa25 please reply to this..

Comment (rtpa25) [reply test/unittesting/testingutils.go:434]: Hey it is actually being used inside authFlow_Test file, it is being used in the emailExistsAndVerificationCheck_test file, it is also being used in the session_test file

File: recipe/session/testingUtils.go (hunk 23 above)
Comment (rtpa25) [reply recipe/session/testingUtils.go:1]: It is because of the function called resetAll if you want to expose resetAll from inside the unittesting package then you need to import all the recipes inside that package that will cause a cyclical import because all recipes need the unittesting package aswell

File: supertokens/querier.go (hunk 24 above)
Comment (rtpa25) [reply supertokens/querier.go:282]: https://pkg.go.dev/fmt#Errorf
it creates an error and returns the error you can see the function signature in the above link

File: recipe/session/config_test.go (hunk 18 above)
Comment (rtpa25) [reply recipe/session/config_test.go]: because I have checked all of them

File: recipe/session/testingUtils.go (hunk 23 above)
Comment (rishabhpoddar) [reply recipe/session/testingUtils.go:1]: cool

**Review by rishabhpoddar — APPROVED** [review]

**Comment by rishabhpoddar** [comment]
Please update changelog and rerun tests, since you branch was out of date with the target branch.

File: CHANGELOG.md [hunk 27]
```diff
@@ -27,6 +27,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ### Change:
 
 -   Uses recipe interface inside session class so that any modification to those get reflected in the session class functions too.
+-   Adds testing framework along with unit tests for the recipes
```
Comment (rishabhpoddar) [inline CHANGELOG.md:30]: This should be under the unreleased section.. since it's not released yet.

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]
